### ISSUE TITLE
Failing case for paper: No related papers; grounding issue

### ISSUE THREAD

**Issue by jvwong** [PathwayCommons/factoid#965]
I cannot seem to get any related papers for a particular article at either the document or network level. There are also some entity grounding display issues, with organisms included when they shouldn't be or missing where you would expect them to be, as described below. These might possibly be related.

Reproduce failing case: 
  - PMID: 32857952
  - Interaction: Guanosine pentaphosphate (CHEBI:16690)/Gsk (NCBI Gene 914634)

B. No related papers

1. INDRA error
 getRelPprsForDoc: FetchError: invalid json response body at https://semanticsearch.baderlab.org/ reason: Unexpected token I in JSON at position 0

A. Grounding display issue:

1. Organism with ChEBI entry

<img src="https://user-images.githubusercontent.com/4706307/114901014-63f36f80-9de2-11eb-9e2b-78dec925843f.png"  height="350px" />

2. Organism is 'Other' for e.Coli gene

<img src="https://user-images.githubusercontent.com/4706307/114901075-71105e80-9de2-11eb-9122-707b1a94b36a.png"  height="350px" />

**Comment by jvwong** [comment]
Oddly enough, the related papers are updated from via the admin.

**Comment by jvwong** [comment]
Grounding organism for chemical also shows up in explorer: 

![Screenshot_2021-04-15 ppGpp Coordinates Nucleotide and Amino-Acid Synthesis in E coli During Starvation (4)](https://user-images.githubusercontent.com/4706307/114903502-df562080-9de4-11eb-98a9-5185edff9d8c.png)

**Comment by jvwong** [comment]
Not sure if this is related issue but:

The user name/input  is assigned to the name of the top search hit chemical:

e.g. search for `18S ribosomal RNA`

![Screenshot_2021-04-16 Biofactoid(1)](https://user-images.githubusercontent.com/4706307/115037679-bb561600-9e9c-11eb-9e98-454d8cf0ae83.png)

**Comment by maxkfranz** [comment]
Have you verified the index, and have you made sure you're not using an outdated version of the system?

E.g. (latest):
<img width="439" alt="Screen Shot 2021-04-16 at 10 41 16 AM" src="https://user-images.githubusercontent.com/989043/115041340-64ead680-9ea0-11eb-8abf-a48ff61ff7e5.png">

It looks like master hasn't been updated from unstable in quite some time, and the newest docker image is fairly old.

**Comment by jvwong** [comment]
I ran this locally:

- grounding-search (master, Thu Feb 18)
- factoid (unstable, Wed Apr 21)

and I get the same problem with the title: 

<img width="432" alt="image" src="https://user-images.githubusercontent.com/4706307/115775362-cb7c6280-a380-11eb-845c-86bdd7d0fa44.png">

**Comment by jvwong** [comment]
All of a sudden, this grounding-related error has disappeared? Maybe something in my browser(s) - this was happening on the unstable instance. Anyways, the related papers errors remain

**Comment by maxkfranz** [comment]
I haven’t been able to reproduce these issues (certainly grounding issues), so there may be confounding variables. Maybe browser extensions?

As long as we can reliably reproduce these issues, we probably should be able to resolve them. Extension-specific issues, for instance, would be nice to solve if they’re for common extensions — though generally lower priority

**Comment by jvwong** [comment]
I still think that the top hit grounding 'name' should be included in the display. I think currently, that field is overwritten by the text input value. e.g. "asdasdasd" will be the name, but I think the grounding above it would be 'alpha-spinasterol'

**Comment by jvwong** [comment]
Well now I can't reproduce any of the errors. Possibly some external service errors, but I'll keep an eye out.

**Comment by jvwong** [comment]
The text input value is used as the name/title of the top hit (overwrites the database record name). I think we still need to show the top hit name. This is confusing when the grounding service provides  an obvious miss. 

For example: https://github.com/PathwayCommons/factoid/issues/965#issuecomment-821209140

**Comment by jvwong** [comment]
This issue of interest has gotten muddled here. I'm reopening a fresh one to explain my problem.

**Comment by jvwong** [comment]
> Well now I can't reproduce any of the errors. Possibly some external service errors, but I'll keep an eye out.

OK now its back, on the prod instance input by a real user [DNA-guided genome editing using the Natronobacterium gregoryi Argonaute](https://biofactoid.org/document/74f186e6-b6c9-4938-a56a-b4dc37908782)

Observation: The user ground ["DNA"](https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:16991) but its a ChEBI ground (i.e. no species) yet it shows Drosophila, same with the other node (which the user 'back-doored' and selected an organism that isn't supported)

<img width="999" alt="image" src="https://user-images.githubusercontent.com/4706307/120544636-a33e4580-c3bb-11eb-9881-2f5be425ec0b.png">


﻿

**Comment by maxkfranz** [comment]
Has this been isolated as a service issue (e.g. poor indexing) or an app-specific issue?

**Comment by jvwong** [comment]
This bug is very elusive:

I tried to reproduce it on biofactoid.org:

  1. Enter title: `DNA-guided genome editing using the Natronobacterium gregoryi Argonaute.`
  2. Add node labelled 'Dishevelled'
    - Click off node to close tooltip
  3. Click same node labelled 'Dishevelled'
    - Click 'Select a better match'
    - Paste in label `Argonaute (Natronobacterium gregoryi)`

Now the label says: Argonaute (Natronobacterium gregoryi) (Drosophila melanogaster)

I try the exact same sequence again in the same/fresh document, and I can't reproduce the error?????

**Comment by maxkfranz** [comment]
When you enter those queries directly into the search service itself (e.g. via insomnia), do you get the expected results?

**Comment by jvwong** [comment]
Yes the results look fine directly. I think I even tried this with different browsers and incognito etc just to make sure it wasn't something local. Strange

**Comment by maxkfranz** [comment]
What about production builds versus local builds?

Is is reproducible only on production, or is it also reproducible on local instances?

**Comment by jvwong** [comment]
I'm seeing this still on production, with recent contributions:

e.g. https://biofactoid.org/document/047c96a4-1006-4052-aa97-9e42ed8e6a9e

![PDCD2-functions-as-an-evolutionarily-conserved-chaperone-dedicated-for-the-40S-ribosomal-protein-uS5-RPS2-Biofactoid](https://user-images.githubusercontent.com/4706307/128757385-49e63d8c-ccb5-44a7-b4aa-47e04f6f2d02.png)

**Comment by maxkfranz** [comment]
You've established that this is reproducible on prod.  Is this also reproducible on a local dev instance?

I haven’t been able to reproduce this issue in either case.  This issue seems difficult to diagnose, so we should narrow down the possibilities.  For instance, is this is a build issue or a business logic issue?  If it’s reproducible on both dev and prod, then it’s probably not a build issue.

Do you have steps that reproduce the issue 100% of the time?

**Comment by jvwong** [comment]
Here's how I can reproduce these results.

1. Entity "Nascent RPS2"

- Add entity and label "RPS2"
- Ground the entity to ["Ribosomal protein S2 [ Drosophila melanogaster (fruit fly) ]"](https://www.ncbi.nlm.nih.gov/gene/34309)
- Edit the label to read "Nascent RPS2" and accept the top ground from ChEBI [Trp-Ser](https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:73694)
- Inspect the associated element (via rethink admin; /api)
 
Here, there seems to the correct ground for ChEBI "Trp-Ser" but this is polluted with stuff from Fly:

```json
"elements": [

    {
        "association": {
            "charge": 0 ,
            "combinedOrganismIndex": 0 ,
            "dbName": "ChEBI" ,
            "dbPrefix": "CHEBI" ,
            "dbXrefs": [
                {
                    "db": "FLYBASE" ,
                    "id": "FBgn0004867"
                }
            ] ,
            ...
            "name": "Trp-Ser" ,
            "nameDistance": 0.6 ,
            "namespace": "chebi" ,
            "organism": "7227" ,
            "organismIndex": 0 ,
            "organismName": "Drosophila melanogaster" ,
            "overallDistance": 6000000060 ,
           ...
            "type": "chemical"
        } ,
       ...
        "name": "Nascent RPS2" ,
       ...
        "type": "chemical"
    }

] , 
```  


2.  Entity "Egfr"

- Add entity and label "Egfr" and accept the top ground ["epidermal growth factor receptor [ Homo sapiens (human) ]"](https://www.ncbi.nlm.nih.gov/gene/1956)
- Add another entity labelled "Egfr" but ground to the second selection ["epidermal growth factor receptor [ Mus musculus (house mouse) ]"](https://www.ncbi.nlm.nih.gov/gene/13649)
- Inspect the associated element (via rethink admin; /api)

Here, the info for the mouse Egfr (but not the human one) gets polluted with some chemical info for what appears to be [glutamic acid](https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:18237):

```json
{

    "association": {
        "charge": 0 ,
        "combinedOrganismIndex": 4 ,
        "dbName": "NCBI Gene" ,
        "dbPrefix": "ncbigene" ,
        "dbXrefs": [
            {
                "db": "MGI" ,
                "id": "MGI:95294"
            } ,
            {
                "db": "Ensembl" ,
                "id": "ENSMUSG00000020122"
            }
        ] ,
        "defaultOrganismIndex": 3 ,
        "distance": 0 ,
        "esScore": 10.96944 ,
        "formulae": [
            "C5H9NO4"
        ] ,
        "id": "13649" ,
        "inchi": "InChI=1S/C5H9NO4/c6-3(5(9)10)1-2-4(7)8/h3H,1-2,6H2,(H,7,8)(H,9,10)/t3-/m0/s1" ,
        "inchiKey": "WHUUTDBJXJRKMK-VKHMYHEASA-N" ,
        "mass": 147.1293 ,
        "monoisotopicMass": 147.05316 ,
        "name": "Egfr" ,
        "nameDistance": 0 ,
        "namespace": "ncbi" ,
        "organism": "10090" ,
        "organismIndex": 1 ,
        "organismName": "Mus musculus" ,
        "overallDistance": 400000 ,
        "shortSynonyms": [
            "9030024J15Rik" ,
            "AI552599" ,
            "Erb" ,
            "Erbb" ,
            "Err" ,
            "epidermal growth factor receptor" ,
            "Wa" ,
            "Errb1" ,
            "Wa5" ,
            "wa2"
        ] ,
        "summary": "An optically active form of glutamic acid having L-configuration." ,
           ...
        "type": "protein"
    } ,
       ...
    "name": "Egfr" ,
   ...
    "type": "protein"

}
```

Now strangely, if you repeat #2 with the label `p53` (human, then mouse), everything is fine????

**Comment by jvwong** [comment]
To consider:

https://github.com/PathwayCommons/factoid/blob/ea906734773f74d56d6922df4735465987246055/src/model/syncher.js#L717

My best guess is that rethinkdb `.update( newAssociation )` effectively merges in changes to the existing rethink Doc and since there are entity association fields that differ between genes and chemicals (e.g. inchi, organismName) you'll get those hanging around?

**Comment by maxkfranz** [comment]
That's a good observation.  Unfortunately, there's no way to use `.replace()` instead of `.update()` here, as that API is for an entire document rather than a field in RDB.  We have to continue using the `.update()` approach in the line you mentioned in order for the writes to work (esp. for them to be atomic).

You outlined the reproducible steps.  The next step is to have automated tests for those steps.  I've approximated them in 782a320.  Note that they're failing on the server side writes (i.e. probably related to writing to the DB).

The changes in 46532c5 resolve those grounding tests not by modifying the synching architecture -- as that would be risky and difficult -- but by amending the grounding `associate()` method to explicitly request to clear out fields that aren't being overwritten by the new grounding.

Could you try that out and confirm whether those changes also affect the UI for your steps?  It may also make sense to make the test cases use real data.  I've put in simple placeholder fields for testing.  If you could insert your example grounding JSON into those test cases, that would help to increase our confidence whether or not we have a fix.

You can run these tests quickly (much more quickly than the UI) with this:

```
npx mocha -r esm -g "associates with grounding"
```

**Comment by jvwong** [comment]
Can confirm those eliminate the problems in the UI. 

Will update tests with grounding data....

**Comment by jvwong** [comment]
OK I think afbd378923a82070a5eb79d0221fbcb930bb62ad should do it.

**Comment by maxkfranz** [comment]
Some final details to consider:

- We may need to copy/clone your objects (maybe via `_.clone()`), otherwise the tests may pass just because multiple objects are pointing to the same grounding object in memory.
-  Are there any issues with nested objects (arrays or objects), like the synonyms, or do they get completely overwritten by the second grounding?

**Comment by jvwong** [comment]
>* We may need to copy/clone your objects (maybe via `_.clone()`), otherwise the tests may pass just because multiple objects are pointing to the same grounding object in memory.

Like this ? 

```js
it('associates with grounding x2 on self client', function( done ){
      const grd_gpp = _.cloneDeep( TEST_GROUND_GGP );
      const grd_chem = _.cloneDeep( TEST_GROUND_CHEMICAL );

      eleC1.associate( grd_gpp );

      eleC1.associate( grd_chem );

      // old grounding should be gone from client 1
      expect( eleC1.association().organism ).to.not.exist;

      // new grounding should be there for client 1
      expect( eleC1.association().monoisotopicMass ).to.equal( TEST_GROUND_CHEMICAL.monoisotopicMass );

      done();
    });

```

**Comment by maxkfranz** [comment]
Exactly

**Comment by jvwong** [comment]
>     * Are there any issues with nested objects (arrays or objects), like the synonyms, or do they get completely overwritten by the second grounding?

I just console log'd the tests and they overwrite objects, so seems OK...

**Comment by jvwong** [comment]
I included a simple check for the arrays: 66958cf4a70712ccc2f31d462bbbc88786414491

**Comment by maxkfranz** [comment]
Just wanted to put it out there, since the synonyms would be a subtle possibility for error.  Sounds like we’re good to go
